tests/share/models/test_cascade.py

```diff
@@ -0,0 +1,300 @@
+
+    def test_many_versions(self):
+        cw = factories.AbstractCreativeWorkFactory(title='All about cats')
+        cw.administrative_change(title='All About Kets')
+
+        assert models.CreativeWork.objects.all().count() == 1
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 2
+
+        assert cw.versions.count() == 2
+
+        cw.delete()
+
+        assert models.CreativeWork.objects.all().count() == 0
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 0
+
+    def test_multiple_sources(self):
+        pass
+
+    def test_no_truncate(self):
+        cws = [factories.AbstractCreativeWorkFactory() for _ in range(10)]
+
+        assert models.CreativeWork.objects.all().count() == 10
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 10
+
+        for i, cw in enumerate(cws):
+            cw.delete()
+            assert models.CreativeWork.objects.all().count() == 9 - i
+            assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 9 - i
+
+
+@pytest.mark.django_db
+class TestDeleteCascadeRelations:
+
+    def test_foreign_key(self):
+        identifier = factories.WorkIdentifierFactory()
+
+        assert models.WorkIdentifier.objects.exclude(id=None).count() == 1
+        assert models.AbstractCreativeWork.objects.exclude(id=None).count() == 1
+        assert models.WorkIdentifierVersion.objects.exclude(id=None).count() == 1
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 1
+
+        identifier.delete()
+
+        assert models.WorkIdentifier.objects.exclude(id=None).count() == 0
+        assert models.AbstractCreativeWork.objects.exclude(id=None).count() == 1
+        assert models.WorkIdentifierVersion.objects.exclude(id=None).count() == 0
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 1
+
+    def test_foreign_key_inverse(self):
+        identifier = factories.WorkIdentifierFactory()
+
+        assert models.WorkIdentifier.objects.exclude(id=None).count() == 1
+        assert models.AbstractCreativeWork.objects.exclude(id=None).count() == 1
+        assert models.WorkIdentifierVersion.objects.exclude(id=None).count() == 1
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 1
+
+        identifier.creative_work.delete()
+
+        assert models.WorkIdentifier.objects.exclude(id=None).count() == 0
+        assert models.AbstractCreativeWork.objects.exclude(id=None).count() == 0
+        assert models.WorkIdentifierVersion.objects.exclude(id=None).count() == 0
+        assert models.AbstractCreativeWorkVersion.objects.exclude(id=None).count() == 0
+
+    def test_many_to_many(self):
+        work = factories.AbstractCreativeWorkFactory()
+        for _ in range(10):
+            factories.ThroughTagsFactory(creative_work=work)
```

Comment:
This morning I got around to checking your branch out and testing locally -- looks like this factory sometimes makes duplicate tags:
```
django.db.utils.IntegrityError: duplicate key value violates unique constraint "share_tag_name_key"
DETAIL:  Key (name)=(corrupti) already exists.